src/compiler/tracing.ts

```diff
@@ -245,17 +284,26 @@ namespace ts { // eslint-disable-line one-namespace-per-file
                     intrinsicName: (type as any).intrinsicName,
                     symbolName: symbol?.escapedName && unescapeLeadingUnderscores(symbol.escapedName),
                     recursionId: recursionToken,
+                    isTuple: objectFlags & ObjectFlags.Tuple ? true : undefined,
                     unionTypes: (type.flags & TypeFlags.Union) ? (type as UnionType).types?.map(t => t.id) : undefined,
                     intersectionTypes: (type.flags & TypeFlags.Intersection) ? (type as IntersectionType).types.map(t => t.id) : undefined,
                     aliasTypeArguments: type.aliasTypeArguments?.map(t => t.id),
                     keyofType: (type.flags & TypeFlags.Index) ? (type as IndexType).type?.id : undefined,
                     ...indexedAccessProperties,
                     ...referenceProperties,
                     ...conditionalProperties,
-                    firstDeclaration: firstDeclaration && {
+                    ...substitutionProperties,
+                    ...reverseMappedProperties,
+                    ...evolvingArrayProperties,
+                    destructuringPattern: destructuringPatternFile && {
+                        path: destructuringPatternFile.path,
+                        start: indexFromOne(getLineAndCharacterOfPosition(destructuringPatternFile, destructuringPattern!.pos)),
+                        end: indexFromOne(getLineAndCharacterOfPosition(destructuringPatternFile, destructuringPattern!.end)),
+                    },
+                    firstDeclaration: firstFile && {
                         path: firstFile.path,
                         start: indexFromOne(getLineAndCharacterOfPosition(firstFile, firstDeclaration.pos)),
-                        end: indexFromOne(getLineAndCharacterOfPosition(getSourceFileOfNode(firstDeclaration), firstDeclaration.end)),
+                        end: indexFromOne(getLineAndCharacterOfPosition(firstFile, firstDeclaration.end)),

```

Comment:
Since this is used much more now, and it's always `indexFromOne(getLineAndCharacterOfPosition(...))`, then maybe make `indexFromOne` do the `getLineAndCharacterOfPosition` call?

Reply:
Pulled out a helper for all three callers.

Reply:
Nice!